lib/ProductOpener/APIProductServices.pm

```diff
@@ -178,57 +208,82 @@ Process API v3 product services requests.
+            my $service_function = $service_functions{$service};
+
+            if (defined $service_function) {
+                if ($service eq 'check_quality') {
+                    # Create a temporary product reference for quality checks
+                    my $temp_product_ref = {};
+                    $temp_product_ref->{nutrition} = $request_body_ref->{nutrition} if defined $request_body_ref->{nutrition};
+                    $temp_product_ref->{ingredients} = $request_body_ref->{ingredients} if defined $request_body_ref->{ingredients};
+
+                    # Call the check_quality service, passing the temporary product ref
+                    &$service_function($temp_product_ref, $request_ref->{updated_product_fields});
+                    
+                    # Integrate the quality check results back into the main product_ref
+                    $product_ref->{data_quality_tags} = $temp_product_ref->{data_quality_tags} if defined $temp_product_ref->{data_quality_tags};
+                } else {
+                    &$service_function($product_ref, $request_ref->{updated_product_fields});
+                }
+            } else {
+                add_error(
+                    $response_ref,
+                    {
+                        message => {id => "unknown_service"},
+                        field => {id => "services", value => $service},
+                        impact => {id => "failure"},
+                    }
+                );
+            }
+        }
+
+        # Select / compute only the fields requested by the caller, default to updated fields
+        my $fields_ref = request_param($request_ref, 'fields');
+        if (not defined $fields_ref) {
+            $fields_ref = ["updated"];
+        }
+        $log->debug("product_services_api - before customize", {fields_ref => $fields_ref, product_ref => $product_ref})
+            if $log->is_debug();
+        $response_ref->{product} = customize_response_for_product($request_ref, $product_ref, undef, $fields_ref);
+
+        # Echo back the services that were executed
+        $response_ref->{fields} = $fields_ref;
+    }
+
+    $log->debug("product_services_api - stop", {request => $request_ref}) if $log->is_debug();
+
+    return;
+}
 
-	return;
+sub check_quality_service ($product_ref, $updated_product_fields_ref) {
+    # Check if nutrition data is provided
+    if (exists $product_ref->{nutrition}) {
+        ProductOpener::DataQuality::check_quality($product_ref->{nutrition});
+        $updated_product_fields_ref->{nutrition_data_quality_tags} = $product_ref->{nutrition}->{data_quality_tags};
+    }
+
+    # Check if ingredient data is provided
+    if (exists $product_ref->{ingredients}) {
+        ProductOpener::DataQuality::check_quality($product_ref->{ingredients});
+        $updated_product_fields_ref->{ingredient_data_quality_tags} = $product_ref->{ingredients}->{data_quality_tags};
+    }
```

Comment:
check_quality needs the product_ref directly not the field.

Reply:
I would call check_quality($product_ref) and then remove eventual false positive (warnings that come from missing fields).

Maybe we will have to categorize the quality errors/warning in respective taxonomies to say on which field they are dependent to filter them.

Or maybe we should split the check_quality function with more granularity or even add a parameter to say which fields we want to check quality.